crates/emmylua_parser_desc/src/util.rs

```diff
@@ -0,0 +1,504 @@
+            if start > cursor_position {
+                lines.truncate(i);
+                break;
+            }
+        }
+    }
+
+    lines
+}
+
+pub trait ResultContainer {
+    fn results(&self) -> &Vec<DescItem>;
+
+    fn results_mut(&mut self) -> &mut Vec<DescItem>;
+
+    fn cursor_position(&self) -> Option<usize>;
+
+    fn emit_range(&mut self, range: SourceRange, kind: DescItemKind) {
+        let should_emit = if let Some(cursor_position) = self.cursor_position() {
+            kind == DescItemKind::Ref && range.contains_inclusive(cursor_position)
+        } else {
+            !range.is_empty()
+        };
+
+        if should_emit {
+            let Some(last) = self.results_mut().last_mut() else {
+                self.results_mut().push(DescItem {
+                    range: range.into(),
+                    kind,
+                });
+                return;
+            };
+
+            let end: usize = last.range.end().into();
+            if last.kind == kind && end == range.start_offset {
+                last.range = last.range.cover(range.into());
+            } else {
+                self.results_mut().push(DescItem {
+                    range: range.into(),
+                    kind,
+                });
+            }
+        }
+    }
+
+    fn emit(&mut self, reader: &mut Reader, kind: DescItemKind) {
+        self.emit_range(reader.current_range(), kind);
+        reader.reset_buff();
+    }
+}
+
+pub struct BacktrackPoint<'a> {
+    prev_reader: Reader<'a>,
+    prev_pos: usize,
+}
+
+impl<'a> BacktrackPoint<'a> {
+    pub fn new<C: ResultContainer>(c: &mut C, reader: &mut Reader<'a>) -> Self {
+        Self {
+            prev_reader: reader.clone(),
+            prev_pos: c.results().len(),
+        }
+    }
+
+    pub fn commit<C: ResultContainer>(self, c: &mut C, reader: &mut Reader<'a>) {
+        let (_c, _reader) = (c, reader); // We don't actually do anything.
+        std::mem::forget(self);
+    }
+
+    pub fn rollback<C: ResultContainer>(self, c: &mut C, reader: &mut Reader<'a>) {
+        *reader = self.prev_reader.clone();
+        c.results_mut().truncate(self.prev_pos);
+        std::mem::forget(self);
+    }
+}
+
+impl<'a> Drop for BacktrackPoint<'a> {
+    fn drop(&mut self) {
+        panic!("backtrack point should be committed or rolled back");
+    }
+}
+
+pub fn is_punct(c: char) -> bool {
+    if c.is_ascii() {
+        c.is_ascii_punctuation()
+    } else {
+        static RE: LazyLock<regex::Regex> =
+            LazyLock::new(|| regex::Regex::new(r"^\p{P}|\p{S}$").unwrap());
+        let mut tmp = [0; 4];
+        RE.is_match(c.encode_utf8(&mut tmp))
+    }
+}
+
+pub fn is_opening_quote(c: char) -> bool {
+    if c.is_ascii() {
+        matches!(c, '\'' | '"' | '<' | '(' | '[' | '{')
+    } else {
+        static RE: LazyLock<regex::Regex> =
+            LazyLock::new(|| regex::Regex::new(r"^\p{Ps}|\p{Pi}|\p{Pf}$").unwrap());
+        let mut tmp = [0; 4];
```

Comment:
I don’t want to introduce regular expressions into the base library. Maybe you can use unicode-general-category instead.